service/topology_coordinator.cc

```diff
@@ -1068,8 +1068,29 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                             .build());
                     }
                     break;
-                case locator::tablet_transition_stage::write_both_read_new:
-                    transition_to_with_barrier(locator::tablet_transition_stage::use_new);
+                case locator::tablet_transition_stage::write_both_read_new: {
+                    auto next_stage = locator::tablet_transition_stage::use_new;
+                    if (action_failed(tablet_state.barriers[trinfo.stage])) {
+                        auto tsi = get_migration_streaming_info(get_token_metadata().get_topology(), tmap.get_tablet_info(gid.tablet), trinfo);
```

Comment:
We should not be counting replicas from streaming info, but compare dead count in tinfo.replicas vs trinfo.next. Because we're interested in which replica set has higher availability. That also depends on the state of nodes which we didn't stream from.

Reply:
Agree. However, this raises one more question.

Let's assume the tinfo.replicas = {A, B, C} and tinfo.new_replicas = {A, B, D}, i.e. streaming C->D happened. At this stage streaming is done and we have tablet copy available at {A, B, C, D}. Now if A is dead and removed, write_both_read_new fails, and whatever counting we apply here we'll end up in either {A, B, C} or {A, B, D}, but {A, B, C, D} looks much better.

Reply:
> but {A, B, C, D} looks much better.

Why? Note that we cannot stay there, and availability is not improved in this state since CL for writes is increased so writes require {B, C, D} for quorum and reads are already using {A, B, D}. We could theoretically switch to {B, C, D} in this case, but I'm not sure it's worth the complexity.

Reply:
> Why? 

Because if we cleanup tablet on either D or C, we will need to copy it again later to compensate for lost copy on A.

>  We could theoretically switch to {B, C, D} in this case, but I'm not sure it's worth the complexity.

Yes, that's indeed extra logic we may want to avoid